etcdserver/api/membership/cluster.go

```diff
@@ -252,6 +252,16 @@ func (c *RaftCluster) Recover(onSet func(*zap.Logger, *semver.Version)) {
 	}
 }
 
+// IsPromoteChange checks if m is a promoteChange
+func (c *RaftCluster) IsPromoteChange(m *Member) bool {
+	members, _ := membersFromStore(c.lg, c.v2store)
```

Comment:
why do we need to load member from c.v2store, not from c.members directly?

Reply:
Members info from store is more "accurate" then c.members. In most cases they are the same, except during server bootstrapping.

During server bootstrapping, we construct members information from "--init-cluster". The resulting members info does not have any information on member's learner flag. Because we do not provide a learner flag when starting etcd binary.

Also check here, in current code we load members from store when validating the conf change in apply stage, it is for very similar reasons.
https://github.com/etcd-io/etcd/blob/a44a281ac3378c4e050e08b1c45c500c731a78eb/etcdserver/api/membership/cluster.go#L255-L256

Reply:
Maybe I should be more specific on this topic. Consider the case where we add a new member to an existing cluster. 
1. Call MemberAdd API to add new member to existing members in cluster. 
 1.1 Each existing member append confchange raft entry of adding the new member.
 1.2 Each existing member apply the raft entry which adds the new member to its `c.members` and backend storage.
2. When starting the new member (during bootstrapping), new member's `c.members` is initially constructed from the input parameter "--init-cluster", whereas initally the backend storage has no members information. At this point, new member's `c.members` is almost accurate except that it does not have info on learner.
3. As the new member getting raft entries from leader and applying them, it will eventually apply all the confchange raft entries, including the entry that adds itself to the cluster. At this point, its `c.members` is accurate.

During bootstrapping, apply stage should always refer to member info form backend storage. Otherwise server gets confused when applying the confchange raft entries: adding a member which already exists in its `c.members`.

Reply:
BTW, just realized this function is not part of the final version. We had to come up with a better way of deciding (in apply stage) if a confchange is for adding new member, or for promoting an existing member.
https://github.com/jingyih/etcd/blob/b433162c0b9ea4d13e9a0dd49b1f51e20896cbcc/etcdserver/server.go#L2232

Reply:
ok. then probably replace this part of the PR with the final version :)

Reply:
It will be removed in later commits. You will not see this function in part2 PR.

Reply:
thanks!